eng/scripts/Pack-Crates.ps1

```diff
@@ -0,0 +1,330 @@
+    }
+  }
+}
+
+function Get-ApiMetadata($package) {
+  $packagePath = Split-Path -Path $package.manifest_path -Parent
+  $readmePath = Join-Path -Path $packagePath -ChildPath $package.readme
+
+  $jsonBody = [ordered]@{
+    'name'          = $package.name
+    'vers'          = $package.version
+    'deps'          = @()
+    'features'      = $package.features
+    'authors'       = $package.authors
+    'description'   = $package.description
+    'documentation' = $package.documentation
+    'homepage'      = $package.homepage
+    'readme'        = if (Test-Path -Path $readmePath) {
+      Get-Content -Path $readmePath -Raw
+    }
+    else {
+      $null
+    }
+    'readme_file'   = $package.readme
+    'keywords'      = $package.keywords
+    'categories'    = $package.categories
+    'license'       = $package.license
+    'license_file'  = $package.license_file
+    'repository'    = $package.repository
+    'links'         = $package.links
+    'rust_version'  = $package.rust_version
+  }
+
+  foreach ($dependency in $package.dependencies) {
+    $jsonBody.deps += @{
+      'name'                  = $dependency.name
+      'version_req'           = $dependency.req
+      'features'              = $dependency.features
+      'optional'              = $dependency.optional
+      'default_features'      = $dependency.default_features
+      'target'                = $dependency.target
+      'kind'                  = $dependency.kind
+      'explicit_name_in_toml' = $dependency.rename
+    }
+  }
+
+  return $jsonBody
+}
+
+function New-ApiPutFile($crateMetadata, $CrateFilePath) {
+  $metadataBytes = [Text.Encoding]::Utf8.GetBytes($crateMetadata)
+  $metadataLengthBytes = [BitConverter]::GetBytes([UInt32]$metadataBytes.Length)
+  $crateBytes = [IO.File]::ReadAllBytes($CrateFilePath)
+  $crateLengthBytes = [BitConverter]::GetBytes([UInt32]$crateBytes.Length)
+
+  $bytes += $metadataLengthBytes + $metadataBytes + $crateLengthBytes + $crateBytes
+
+  return $bytes
+}
+
+Push-Location $RepoRoot
+try {
+  $localRegistryPath = Initialize-LocalRegistry
+
+  [array]$packages = Get-PackagesToBuild
+
+  Add-PathVersions $packages
+
+  Write-Host "Building packages in the following order:"
+  foreach ($package in $packages) {
+    $packageName = $package.name
+    $type = if ($package.OutputPackage) { "output" } else { "dependency" }
+    Write-Host "  $packageName ($type)"
+  }
+  Write-Host ""
+
+  foreach ($package in $packages) {
+    $packageName = $package.name
+    $packageVersion = $package.version
+
+    Invoke-LoggedCommand `
+      -GroupOutput `
+      -Command "cargo package --package $packageName --config `"source.crates-io.replace-with='local'`" --config `"source.local.local-registry='$localRegistryPath'`" --allow-dirty"
+
+    $crateFile = "$RepoRoot/target/package/$packageName-$packageVersion.crate"
+
+    # copy the package to the local registry
+    Add-CrateToLocalRegistry `
```

Comment:
Probably should double check that it exists. Not sure how well-documented this path is and it could change, or does `Add-CrateToLocalRegistry` suitably fail with an obvious message if it doesn't exist?

Reply:
thanks.  I added a test-path sanity check after the assignment